File: Codeforces/Div 3/Codeforces Round 991/F_Maximum_modulo_equality.cpp
/*
    同餘定理 + 線段樹

    若兩個數A、B ，滿足 A % M == B % M，則兩個數的差值 (A - B) % M == 0，即 M 是 (A - B) 的因數。
    故求最大的 M 等同求兩數之差的最大公因數，可以計算 diffs 陣列，其中 diffs[i] = abs(A[i+1] - A[i])。

    而求區間 [l, r] 的最大 M 值，等同求 diffs[l] ~ diffs[r-1] 的最大公因數，這可以用線段樹維護。
*/

#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

class SegmentTree {
private:
    vector<int> nums;
    vector<int> tree;
public:
    SegmentTree(vector<int>& nums) {
        int n = nums.size();
        this->nums = vector<int>(n + 1, 0);
        for (int i = 1; i <= n; i++) this->nums[i] = nums[i - 1]; // 1-indexed
        this->tree = vector<int>(4 * n, 0);
        build(1, 1, n);
    }

    void build(int o, int left, int right) {
        if (left == right) {
            tree[o] = nums[left];
            return;
        }
        int mid = (left + right) / 2;
        build(2 * o, left, mid);
        build(2 * o + 1, mid + 1, right);
        tree[o] = merge(2 * o, 2 * o + 1);
    }

    int merge(int left_child, int right_child) {
        return gcd(tree[left_child], tree[right_child]);
    }

    int query(int o, int left, int right, int l, int r) {
        if (l <= left && right <= r) return tree[o];
        int mid = (left + right) / 2;
        int ans = 0;
        if (l <= mid) ans = gcd(ans, query(2 * o, left, mid, l, r));
        if (r > mid) ans = gcd(ans, query(2 * o + 1, mid + 1, right, l, r));
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int t, n, q, l, r;
    cin >> t;
    while (t--) {
        cin >> n >> q;
        vector<int> A(n);
        for (int i = 0; i < n; ++i) cin >> A[i];
        vector<int> diffs(n - 1);
        for (int i = 0; i < n - 1; ++i) {
            diffs[i] = abs(A[i + 1] - A[i]);
        }
        SegmentTree* st = (n > 1) ? new SegmentTree(diffs) : nullptr;
        vector<int> ans;
        while (q--) {
            cin >> l >> r;
            if (n == 1 || l == r) ans.push_back(0);
            else ans.push_back(st->query(1, 1, n - 1, l, r - 1));
        }
        for (int x : ans) cout << x << " ";
        cout << endl;
        delete st;
    }
    return 0;
}

File: Codeforces/Div 3/Codeforces Round 991/F_Maximum_modulo_equality.py
"""
    同餘定理 + 線段樹

    若兩個數A、B ，滿足 A % M == B % M，則兩個數的差值 (A - B) % M == 0，即 M 是 (A - B) 的因數。
    故求最大的 M 等同求兩數之差的最大公因數，可以計算 diffs 陣列，其中 diffs[i] = abs(A[i+1] - A[i])。

    而求區間 [l, r] 的最大 M 值，等同求 diffs[l] ~ diffs[r-1] 的最大公因數，這可以用線段樹維護。
"""
from math import gcd

class SegmentTree:
    def __init__(self, nums):
        n = len(nums)
        self.nums = [0] + nums # 讓 index 從 1 開始
        # self.tree = [0 for _ in range(4 * n)]
        self.tree = [0 for _ in range(1 << (n.bit_length() + 1))]
        self.build(1, 1, n)
 
    def build(self, o, left, right): # node, left, right
        if left == right: # Leaf node initialization
            self.tree[o] = self.nums[left]
            return
        mid = (left + right) // 2
        self.build(o << 1, left, mid) # left child
        self.build(o << 1 | 1, mid + 1, right) # right child
        self.tree[o] = self.merge(o << 1, o << 1 | 1)
 
    def merge(self, left_child, right_child):
        return gcd(self.tree[left_child], self.tree[right_child])
 
    def query(self, o, left, right, l, r):
        if l <= left and right <= r:
            return self.tree[o]
        mid = (left + right) // 2
        ans = 0
        if l <= mid:
            ans = gcd(ans, self.query(o << 1, left, mid, l, r))
        if r > mid:
            ans = gcd(ans, self.query(o << 1 | 1, mid + 1, right, l, r))
        return ans

t = int(input())
for _ in range(t):
    n, q = map(int, input().split())
    A = list(map(int, input().split()))
    diffs = [abs(A[i+1] - A[i]) for i in range(n-1)]
    st = SegmentTree(diffs) if n > 1 else None
    ans = []
    for __ in range(q):
        l, r = map(int, input().split())
        if l == r or not diffs:
            ans.append('0')
        else:
            ans.append(st.query(1, 1, n-1, l, r-1))
    print(*ans)